### PR TITLE
Implement SSL_read_ex() and SSL_write_ex() as documented.

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

**Comment by kroeckx** [comment]
This fails many tests, like tests that expect SSL_read() to return >= 0 for success.

**Comment by kroeckx** [comment]
I'm not sure why I get 0 now in the BIO_read()/BIO_write() cases.

**Comment by mattcaswell** [comment]
Perhaps a simpler solution is to have SSL_read() and SSL_read_ex() to both call an internal ssl_read_internal() function, which just returns the existing 0/-1 return codes (confusing as they are). SSL_read_ex() does the conversion of any 0 or -1 return is converted to 0.

**Comment by kroeckx** [comment]
So I pushed a new commit. This doesn't change anything for the ssltest_old case, it still fails. I guess this is caused by the change in ssl_read() to use SSL_read_ex(), turning the -1 into 0 at that point.

**Comment by kroeckx** [comment]
So I used the internal functions now in bio_ssl.c too, and the test suite still works.

File: ssl/ssl_lib.c [hunk 1]
```diff
@@ -1582,6 +1560,37 @@ int SSL_read_ex(SSL *s, void *buf, size_t num, size_t *readbytes)
...
+int SSL_read_ex(SSL *s, void *buf, size_t num, size_t *readbytes)
+{
+    int ret = ssl_read_internal(s, buf, num, readbytes);
+
+    if (ret < 0)
+        ret = 0;
+    return ret;
+}
+
 int SSL_peek(SSL *s, void *buf, int num)
```
Comment (mattcaswell) [inline ssl/ssl_lib.c]: Don't we need to make a similar change to SSL_peek()?

**Comment by kroeckx** [comment]
New commit pushed

File: include/openssl/ssl.h [hunk 2]
```diff
@@ -2176,8 +2176,10 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_SSL_PARSE_SERVERHELLO_USE_SRTP_EXT         311
 # define SSL_F_SSL_PEEK                                   270
 # define SSL_F_SSL_PEEK_EX                                432
+# define SSL_F_SSL_PEEK_INTERNAL                          496
 # define SSL_F_SSL_READ                                   223
 # define SSL_F_SSL_READ_EX                                434
+# define SSL_F_SSL_READ_INTERNAL                          497
```
Comment (levitte) [inline include/openssl/ssl.h]: You need to redo your `make update`, there's currently a number clash:

``` text
$ make update
!! ERROR: SSL function code 497 assigned twice (collision at SSL_F_SSL_READ_INTERNAL)
!! ERROR: SSL function code 496 assigned twice (collision at SSL_F_TLS_CONSTRUCT_CERT_VERIFY)
There were errors, failing...

make: *** [errors] Error 2
```

File: ssl/ssl_locl.h [hunk 3]
```diff
@@ -1854,6 +1854,8 @@ const char *ssl_protocol_to_string(int version);
 
 # ifndef OPENSSL_UNIT_TEST
 
+__owur int ssl_read_internal(SSL *s, void *buf, size_t num, size_t *readbytes);
+__owur int ssl_write_internal(SSL *s, const void *buf, size_t num, size_t *written);
```
Comment (levitte) [inline ssl/ssl_locl.h]: Is there a reason you're not making them static in `ssl_lib.c`?

**Comment by mattcaswell** [comment]
Ping? I'd really like to get these changes in. I am touching code near this area at the moment.

**Comment by levitte** [comment]
> Ping? I'd really like to get these changes in. I am touching code near this area at the moment.

Have a look at my review

**Comment by mattcaswell** [comment]
>> Ping? I'd really like to get these changes in. I am touching code near this area at the moment.
>
> Have a look at my review

My ping was really directed to @kroeckx who owns this PR.

**Comment by kroeckx** [comment]
I forgot all about it, I thought this was already merged ...

File: ssl/ssl_locl.h (hunk 3 above)
Comment (kroeckx) [inline ssl/ssl_locl.h]: Because they're using in bio_ssl.c too?

File: include/openssl/ssl.h (hunk 2 above)
Comment (kroeckx) [inline include/openssl/ssl.h]: That should be fixed now.

File: ssl/ssl_locl.h (hunk 3 above)
Comment (levitte) [inline ssl/ssl_locl.h]: Ah, now I see. I must have scrolled past that difficult a dozen times without noticing. Ok then...

**Comment by kroeckx** [comment]
So can someone please confirm that I can merge this?

**Comment by levitte** [comment]
Yup, please go ahead

**Pull request closed without merging**